utils/mil_ros_tools/online_bagger/nodes/online_bagger.py

```diff
@@ -0,0 +1,332 @@
+        then break out of loop and be called in the callback function to prevent the function
+        from locking up.
+        """
+
+        i = 0
+        while self.successful_subscription_count == 0 and not rospy.is_shutdown():
+            self.subscribe()
+            rospy.sleep(0.1)
+            i = i + 1
+            if i % 1000 == 0:
+                rospy.logdebug('still subscribing!')
+
+    def subscribe(self):
+
+        """
+        Subscribe to the topics defined in the yaml configuration file
+
+        Function checks subscription status True/False of each topic
+        if True: topic has already been sucessfully subscribed to
+        if False: topic still needs to be subscribed to and
+        subscriber will be run.
+
+        Each element in self.subscriber list is a list [topic, Bool]
+        where the Bool tells the current status of the subscriber (sucess/failure).
+
+        Return number of topics that failed subscription
+        """
+
+        for topic in self.subscriber_list:
+            if not topic[1]:
+                msg_class = rostopic.get_topic_class(topic[0])
+                if msg_class[1] is not None:
+                    self.successful_subscription_count = self.successful_subscription_count + 1
+                    rospy.Subscriber(topic[0], msg_class[0],
+                                     lambda msg, _topic=topic[0]: self.bagger_callback(msg, _topic))
+
+                    topic[1] = True  # successful subscription
+
+    def get_topic_duration(self, topic):
+
+        """
+        Return current time duration of topic
+        """
+
+        return self.topic_messages[topic][-1][0] - self.topic_messages[topic][0][0]
+
+    def get_header_time(self, msg):
+
+        """
+        Retrieve header time if available
+        """
+
+        if hasattr(msg, 'header'):
+            return msg.header.stamp
+        else:
+            return rospy.get_rostime()
+
+    def bagger_callback(self, msg, topic):
+
+        """
+        Streaming callback function, stops streaming during bagging process
+        also pops off msgs from dequeue if stream size is greater than specified ram limit
+
+        Stream, callback function does nothing if streaming is not active
+        """
+
+        if not self.streaming:
+            return
+
+        self.iteration_count = self.iteration_count + 1
+        time = self.get_header_time(msg)
+
+        self.topic_messages[topic].append((time, msg))
+
+        time_diff = self.get_topic_duration(topic)
+
+        # verify streaming is popping off and recording topics
+        if self.iteration_count % 100 == 0:
+            rospy.logdebug('time_difference: %s', time_diff.to_sec())
+            rospy.logdebug('topic: %s', topic)
+            rospy.logdebug('topic type: %s', type(msg))
+            rospy.logdebug('number of topic messages: %s', self.get_topic_message_count(topic))
+
+        if time_diff > rospy.Duration(self.stream_time):
```

Comment:
What happened to the idea of having one large ring buffer - that would solve all the problems I think. Dumping from a specific time would be easy since you'd only have to search through one list and keeping the memory limit would be easy since you'd just non-discriminately purge from the end. It would naturally fill up on the small messages and (almost) always stay under the memory limit.